networkx/algorithms/tests/test_dag.py

```diff
@@ -11,36 +12,15 @@ def setUp(self):
         pass
 
     def test_topological_sort1(self):
-        DG = nx.DiGraph()
-        DG.add_edges_from([(1, 2), (1, 3), (2, 3)])
-        assert_equal(nx.topological_sort(DG), [1, 2, 3])
-        assert_equal(nx.topological_sort_recursive(DG), [1, 2, 3])
+        DG = nx.DiGraph([(1, 2), (1, 3), (2, 3)])
 
-        DG.add_edge(3, 2)
-        assert_raises(nx.NetworkXUnfeasible, nx.topological_sort, DG)
-        assert_raises(nx.NetworkXUnfeasible, nx.topological_sort_recursive, DG)
-
-        DG.remove_edge(2, 3)
-        assert_equal(nx.topological_sort(DG), [1, 3, 2])
-        assert_equal(nx.topological_sort_recursive(DG), [1, 3, 2])
-
-    def test_reverse_topological_sort1(self):
-        DG = nx.DiGraph()
-        DG.add_edges_from([(1, 2), (1, 3), (2, 3)])
-        assert_equal(nx.topological_sort(DG, reverse=True), [3, 2, 1])
-        assert_equal(
-            nx.topological_sort_recursive(DG, reverse=True), [3, 2, 1])
+        consume(nx.topological_sort(DG))
 
         DG.add_edge(3, 2)
-        assert_raises(nx.NetworkXUnfeasible,
-                      nx.topological_sort, DG, reverse=True)
-        assert_raises(nx.NetworkXUnfeasible,
-                      nx.topological_sort_recursive, DG, reverse=True)
+        assert_raises(nx.NetworkXUnfeasible, consume, nx.topological_sort(DG))
 
         DG.remove_edge(2, 3)
-        assert_equal(nx.topological_sort(DG, reverse=True), [2, 3, 1])
-        assert_equal(
-            nx.topological_sort_recursive(DG, reverse=True), [2, 3, 1])
+        consume(nx.topological_sort(DG))
```

Comment:
Are you saying that we can't use list(topological_sort(DG)) and get
reproducible results?
I thought the key(node) sorting would remove that issue...

On Fri, Aug 7, 2015 at 7:04 PM, Neil <notifications@github.com> wrote:

> In networkx/algorithms/tests/test_dag.py
> <https://github.com/networkx/networkx/pull/1662#discussion_r36572764>:
>
> >
> >          DG.remove_edge(2, 3)
> > -        assert_equal(nx.topological_sort(DG, reverse=True), [2, 3, 1])
> > -        assert_equal(
> > -            nx.topological_sort_recursive(DG, reverse=True), [2, 3, 1])
> > +        consume(nx.topological_sort(DG))
>
> The consume ensures that there are no exceptions due to cycles.
>
> The old tests worked only by an accident of stable dict iteration in
> CPython. They were never guaranteed to work in Python. They may not work in
> other implementations of Python (IronPython, PyPy, etc.) They may not work
> in newer versions of Python (which use hash randomization) and even newer
> versions (for which it is proposed to make dict an OrderedDict everywhere).
>
> Since the old tests are essentially broken, I have removed everything but
> the no-cycle-verification.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/networkx/networkx/pull/1662/files#r36572764>.
>